File: python/src/bw/services/orchestrator/batch_runner.py
import threading
import time

from chores import chores_pb2


LOCAL_WORKER_NAME = "local-orchestrator"


class BatchRunner:
    def __init__(
        self,
        consumer_pub,
        logger,
    ) -> None:
        self.consumer_pub = consumer_pub
        self.logger = logger
        self.consumer_pub_lock = threading.Lock()

    def run_chore_filter(
        self,
        chores: chores_pb2.Chores,
        person: chores_pb2.PersonAvailability,
    ) -> None:
        filter_id = f"{chores.chores_id}-{person.person_id}"

        self.logger.info(
            "[orchestrator] running local chore filter: "
            "chores_id=%s filter_id=%s person_id=%s chores=%s "
            "available_minutes=%s",
            chores.chores_id,
            filter_id,
            person.person_id,
            len(chores.chores),
            person.available_minutes,
        )

        request = self._build_chore_filter_request(
            filter_id=filter_id,
            chores=chores,
            person=person,
        )

        result = self._process_chore_filter_request(request)

        summary = self._build_summary(
            chores_id=chores.chores_id,
            context=chores.context,
            results=[result],
        )

        self._publish_summary(summary)

    def _build_chore_filter_request(
        self,
        filter_id: str,
        chores: chores_pb2.Chores,
        person: chores_pb2.PersonAvailability,
    ) -> chores_pb2.ChoreFilterRequest:
        request = chores_pb2.ChoreFilterRequest(
            filter_id=filter_id,
            context=chores.context,
        )

        request.chores.CopyFrom(chores)
        request.person.CopyFrom(person)

        return request

    def _process_chore_filter_request(
        self,
        request: chores_pb2.ChoreFilterRequest,
    ) -> chores_pb2.ChoreFilterResult:
        person = request.person
        available_minutes = person.available_minutes

        accepted_chores: list[chores_pb2.Chore] = []
        rejected_chores: list[chores_pb2.Chore] = []

        used_minutes = 0

        for chore in request.chores.chores:
            if used_minutes + chore.estimated_minutes <= available_minutes:
                accepted_chores.append(chore)
                used_minutes += chore.estimated_minutes

                self.logger.info(
                    "[orchestrator] accepted chore locally: "
                    "filter_id=%s person_id=%s chore_id=%s name=%s "
                    "estimated_minutes=%s used_minutes=%s",
                    request.filter_id,
                    person.person_id,
                    chore.chore_id,
                    chore.name,
                    chore.estimated_minutes,
                    used_minutes,
                )

            else:
                rejected_chores.append(chore)

                self.logger.info(
                    "[orchestrator] rejected chore locally: "
                    "filter_id=%s person_id=%s chore_id=%s name=%s "
                    "estimated_minutes=%s used_minutes=%s available_minutes=%s",
                    request.filter_id,
                    person.person_id,
                    chore.chore_id,
                    chore.name,
                    chore.estimated_minutes,
                    used_minutes,
                    available_minutes,
                )

        result = chores_pb2.ChoreFilterResult(
            filter_id=request.filter_id,
            chores_id=request.chores.chores_id,
            person=person,
            used_minutes=used_minutes,
            remaining_minutes=max(available_minutes - used_minutes, 0),
            context=request.context,
        )

        result.accepted_chores.extend(accepted_chores)
        result.rejected_chores.extend(rejected_chores)

        return result

    def _build_summary(
        self,
        chores_id: str,
        context,
        results: list[chores_pb2.ChoreFilterResult],
    ) -> chores_pb2.ChoreFilterSummary:
        total_chores_accepted = sum(
            len(result.accepted_chores)
            for result in results
        )

        summary = chores_pb2.ChoreFilterSummary(
            chores_id=chores_id,
            total_people_evaluated=len(results),
            total_chores_accepted=total_chores_accepted,
            context=context,
        )

        summary.results.extend(results)

        return summary

    def _publish_summary(
        self,
        summary: chores_pb2.ChoreFilterSummary,
    ) -> None:
        with self.consumer_pub_lock:
            self.consumer_pub.put(summary.SerializeToString())

        self.logger.info(
            "[orchestrator] published local chore filter summary: "
            "chores_id=%s people_evaluated=%s chores_accepted=%s",
            summary.chores_id,
            summary.total_people_evaluated,
            summary.total_chores_accepted,
        )

File: rust/services/orchestrator/src/execution/batch_runner.rs
use std::sync::Arc;

use prost::Message;
use rayon::{
    ThreadPool,
    ThreadPoolBuilder,
};
use tokio::sync::oneshot;

use bw_core::{
    config::DynError,
    proto::demo::{
        chores::{
            ChoreFilterRequest, ChoreFilterResult, ChoreFilterSummary, Chores,
            PersonAvailability,
        },
        common::RequestContext,
    },
};

use crate::processing::processor::process_chore_filter_request;

pub struct BatchRunner {
    summary_pub: Arc<zenoh::pubsub::Publisher<'static>>,
    worker_pool: Arc<ThreadPool>,
}

impl Clone for BatchRunner {
    fn clone(&self) -> Self {
        Self {
            summary_pub: Arc::clone(&self.summary_pub),
            worker_pool: Arc::clone(&self.worker_pool),
        }
    }
}

impl BatchRunner {
    pub fn new(
        summary_pub: zenoh::pubsub::Publisher<'static>,
        num_worker_threads: usize,
    ) -> Result<Self, DynError> {
        let worker_pool = ThreadPoolBuilder::new()
            .num_threads(num_worker_threads)
            .thread_name(|index| {
                format!("orchestrator-worker-{index}")
            })
            .build()?;

        Ok(Self {
            summary_pub: Arc::new(summary_pub),
            worker_pool: Arc::new(worker_pool),
        })
    }

    pub async fn run_chore_filter(
        &self,
        chores: Chores,
        person: PersonAvailability,
    ) -> Result<(), DynError> {
        let filter_id = format!("{}-{}", chores.chores_id, person.person_id);

        tracing::info!(
            "[batch-runner] running local chore filter: chores_id={} filter_id={} person_id={} chores={} available_minutes={}",
            chores.chores_id,
            filter_id,
            person.person_id,
            chores.chores.len(),
            person.available_minutes,
        );

        let request = Self::build_chore_filter_request(
            filter_id,
            chores.clone(),
            person,
        );

        let result = self.run_processor_on_worker(request).await?;

        let summary = Self::build_summary(
            chores.chores_id,
            chores.context,
            vec![result],
        );

        self.publish_summary(summary).await
    }

    pub async fn run_chore_filter_window(
        &self,
        chores: Chores,
        people: Vec<PersonAvailability>,
    ) -> Result<(), DynError> {
        tracing::info!(
            "[batch-runner] running local window chore filter: chores_id={} people={} chores={}",
            chores.chores_id,
            people.len(),
            chores.chores.len(),
        );

        let chores_for_worker = chores.clone();
        let worker_pool = Arc::clone(&self.worker_pool);

        let (tx, rx) = oneshot::channel();

        worker_pool.spawn(move || {
            let mut results = Vec::new();

            for person in people {
                let filter_id = format!(
                    "{}-{}",
                    chores_for_worker.chores_id,
                    person.person_id,
                );

                let request = Self::build_chore_filter_request(
                    filter_id,
                    chores_for_worker.clone(),
                    person,
                );

                results.push(process_chore_filter_request(request));
            }

            let _ = tx.send(results);
        });

        let results = rx.await.map_err(|err| {
            format!("window worker failed to return result: {err}")
        })?;

        let summary = Self::build_summary(
            chores.chores_id,
            chores.context,
            results,
        );

        self.publish_summary(summary).await
    }

    async fn run_processor_on_worker(
        &self,
        request: ChoreFilterRequest,
    ) -> Result<ChoreFilterResult, DynError> {
        let worker_pool = Arc::clone(&self.worker_pool);
        let (tx, rx) = oneshot::channel();

        worker_pool.spawn(move || {
            let result = process_chore_filter_request(request);
            let _ = tx.send(result);
        });

        let result = rx.await.map_err(|err| {
            format!("worker failed to return result: {err}")
        })?;

        Ok(result)
    }

    fn build_chore_filter_request(
        filter_id: String,
        chores: Chores,
        person: PersonAvailability,
    ) -> ChoreFilterRequest {
        ChoreFilterRequest {
            filter_id,
            context: chores.context.clone(),
            chores: Some(chores),
            person: Some(person),
        }
    }

    fn build_summary(
        chores_id: String,
        context: Option<RequestContext>,
        results: Vec<ChoreFilterResult>,
    ) -> ChoreFilterSummary {
        let total_chores_accepted = results
            .iter()
            .map(|result| result.accepted_chores.len() as u32)
            .sum();

        ChoreFilterSummary {
            chores_id,
            total_people_evaluated: results.len() as u32,
            total_chores_accepted,
            results,
            context,
        }
    }

    async fn publish_summary(
        &self,
        summary: ChoreFilterSummary,
    ) -> Result<(), DynError> {
        let mut buf = Vec::with_capacity(summary.encoded_len());
        summary.encode(&mut buf)?;

        self.summary_pub.put(buf).await?;

        tracing::info!(
            "[batch-runner] published local chore filter summary: chores_id={} people_evaluated={} chores_accepted={}",
            summary.chores_id,
            summary.total_people_evaluated,
            summary.total_chores_accepted,
        );

        Ok(())
    }
}